### ISSUE TITLE
Our own backends for X11/Wayland

### ISSUE THREAD

**Issue by kekekeks** [AvaloniaUI/Avalonia#1243]
There are way too many limitations and tradeoffs that GTK/GDK have to support both of them, while not allowing us to do things that we want to do. Notable examples are separate-thread rendering (we have to use our own X11 connection to do that) and issues with absolute window positioning on Wayland (Wayland has `xdg_surface_set_window_geometry` which isn't used by GTK for some reason which leads to our popups being broken).

There are some major caveats on that way:
1) System dialogs (open/save file, folder selection, etc). We need to either still load GTK or have them out-of-proc somehow. The out-of-proc way will cause issues with making dialogs modal to the parent window on X11
2) Input methods. Right now we are relying on GTK do support them
3) Client-side window decorations. Some of the wayland shells are requiring applications to draw window frame, title and buttons, we need to support that somehow

**Comment by martindecker** [comment]
Does "Our own backends for X11/Wayland " mean , you plan to write (more) code to interact with X11 ? I think, because Wayland is the future, it is not necessary to support X11 anymore. Keep it simple and only support Wayland. But I am not expert regarding this question, currently I am a WPF developer.

**Comment by kekekeks** [comment]
Wayland haven't yet reached the feature parity with X11 and judging from their mailing list will never do so, because some features aren't implemented *on purpose*. X11 will also be more widespread for a half a decade, at least.

**Comment by kekekeks** [comment]
Shower thought: GTK is *not*, in fact, a "native" UI toolkit for Linux. Some popular distros are using KDE and Qt. So our dependency on GTK isn't free. Our file dialogs also don't look native everywhere.

So we should use libX11/libxcb for UI and separate backend system for file dialogs, that should support:
- GTK2/3
- Qt 3/4/5
- our own managed implementation that doesn't depend on anything

**Comment by kekekeks** [comment]
According to https://lists.freedesktop.org/archives/wayland-devel/2017-September/034967.html wayland doesn't support dpi awareness. Great.

**Comment by jkoritzinsky** [comment]
It looks like there's support on the Wayland side (and in Weston) as of 1.2, but there isn't support in the XWayland front end.

https://github.com/wayland-project/weston/blob/1a2adfedea74ae264eca7d4b8cbe6d1f04e552bf/man/weston.ini.man#L415

**Comment by kekekeks** [comment]
Yep, `wl_surface::set_buffer_scale` seems to be the way of per-window scaling factors

**Comment by iMonZ** [comment]
> Shower thought: GTK is _not_, in fact, a "native" UI toolkit for Linux. Some popular distros are using KDE and Qt. So our dependency on GTK isn't free. Our file dialogs also don't look native everywhere.
> 
> So we should use libX11/libxcb for UI and separate backend system for file dialogs, that should support:
> 
>     * GTK2/3
> 
>     * Qt 3/4/5
> 
>     * our own managed implementation that doesn't depend on anything

So xcb doesn't work on modern systems.
Maybe dropping gtk2 support would help with the problems.
But full Wayland support without X11 or XWayland or xcb should be on the roadmap.
Ubuntu LTS will switch to Wayland in 22.04 and most other popular distros already switched to Wayland.
And most feature aren't implemented in Wayland for security reason, so yes it's on purpose.

**Comment by jgcodes2020** [comment]
> Shower thought: GTK is _not_, in fact, a "native" UI toolkit for Linux. Some popular distros are using KDE and Qt. So our dependency on GTK isn't free. Our file dialogs also don't look native everywhere.
> 
> So we should use libX11/libxcb for UI and separate backend system for file dialogs, that should support:
> 
>     * GTK2/3
> 
>     * Qt 3/4/5
> 
>     * our own managed implementation that doesn't depend on anything

File dialogs can be handled via desktop portals: https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.FileChooser

**Comment by maxkatz6** [comment]
@jgcodes2020 yes, and it was https://github.com/AvaloniaUI/Avalonia/pull/8217

**Comment by microhobby** [comment]
Hey, my dear Avalonia maintainers, let me say that this is important. Some users that are using Avalonia for embedded system does feel more comfortable having a desktop environment than using framebuffer or DRM. So, for embedded some machines does not support X11 anymore, only Wayland and Weston. "XWayland can still be used" may some argue. Yeah, but in some scenarios Xwayland is not supported and also is hard to set openGL ES for Xwayland, so, no HW acceleration for this users...

**Comment by timunie** [comment]
Wayland is planned afaik and the preparation work where we need breaking changes is already in progress. In the 11.x time frame we hope to be able to add wayland support. @affederaffe already did a great job on this backend.

**Comment by iMonZ** [comment]
> Wayland is planned afaik and the preparation work where we need breaking changes is already in progress. In the 11.x time frame we hope to be able to add wayland support. @affederaffe already did a great job on this backend.

Hey 
any news on that?

**Comment by ratin3** [comment]
Gentlemen, I am working  with Zynq / embedded os like Petalinux, was wondering what components I will need to support Avalonia UI and .Net framework that Avalonia is based on. I would like to avoid Xorg if possible. Thanks

**Comment by LoongArum** [comment]
> Gentlemen, I am working with Zynq / embedded os like Petalinux, was wondering what components I will need to support Avalonia UI and .Net framework that Avalonia is based on. I would like to avoid Xorg if possible. Thanks先生们，我正在使用 Zynq / 嵌入式操作系统，例如 Petalinux，我想知道我需要哪些组件来支持 Avalonia 所基于的 Avalonia UI 和 .Net 框架。如果可能的话，我想避免Xorg。谢谢

Just use LinuxFrameBuffer, ther are tutorial in avalonia document.

**Comment by amayra** [comment]
so Wayland support is officially dead ?

thanks gnome devs

**Comment by maxkatz6** [comment]
> so Wayland support is officially dead ?

It's not.

**Comment by jindongjie** [comment]
> > so Wayland support is officially dead ?
> 
> 
> 
> It's not.
So，in future we can use wayland backend without   manually install nuget pkg？

**Comment by Haltroy** [comment]
Can't we just fallback to X11 (XWayland in this case) in GNOME and use Wayland on others that do support what we need?
This way, we can still have a partial support on Wayland until we figure out something for GNOME.
— A KDE Plasma Wayland Session User

**Comment by maxkatz6** [comment]
> So，in future we can use wayland backend without manually install nuget pkg？

That's the plan. We are not working on wayland support right now, but we didn't give up on the idea to support it for 11.x or 12.0 release. 

@Haltroy I would ping @kekekeks or @jmacato to answer that.

**Comment by TheArcaneBrony** [comment]
Would be nice if Wayland support was prioritised, especially given I no longer have a way to reach out for support when things don't work with xwayland.

**Comment by jgcodes2020** [comment]
Given that CSD is the primary blocker for Wayland support, is there a tracker issue or potential API design work that has been done for it?

**Comment by kekekeks** [comment]
The plan is to keep CSD completely hidden from the user code and implement them as a combination of platform backend code and our compositor internal. It would be either a fake TopLevel that renders to the main wl_surface with client area being rendered into wl_subsurface, or some way to overlay multiple TopLevel's into the same composition target instance.

**Comment by Haltroy** [comment]
Wait I'm confused.

CSD ("client-side decorations ") is what GNOME uses which render the title bar (Avalonia's "ExtendClientAreaChromeHints"). SSD (server-side decorations") is what KDE/Qt uses which lets the compositor to draw the title bar. [Source](https://pointieststick.com/2021/05/15/window-decorations-revisited-or-using-the-right-tool-for-the-job/#:~:text=%E2%80%9CClient-side%E2%80%9D%20refers,drawn%20by%20KWin%2C)
So the CSD stuff should be already implemented (since it's a GNOME thing and usually GNOME's things are applied to everyone but not vice-versa which is the case of my comment). The thing that you should use for drawing "ExtendClientAreaChromeHints" should be CSD (in my opinion), correct me  if I'm wrong. If we don't want to extend to titlebar we don't set this value to anything and use the default which is the case where you would use SSD. 

How about this: use CSD completely. Make a default titlebar for Avalonia windows for normal title bar and hide it if this setting is set to a different value. This way you don't have to wait GNOME another millennium to implement SSD to Mutter.

I have no idea how this works or might get it wrong so if I do please correct me. Also is there anything that's blocking Wayland support for Avalonia?

**Comment by kekekeks** [comment]
> is what GNOME uses which render the title bar

GNOME **does not** draw the titlebar, GNOME forces wayland applications to draw their own titlebars. That's not the case for Windows, macOS, X11 and most of Wayland compositors.

Avalonia doesn't currently provide its own titlebar and attempting to put one into Window template will break app expectations of full ClientSize being available to them.

Implementing CSD in a way that's consistent with app expectations on other platforms is doable, but requires work that's currently not planned for the next months.

**Comment by Haltroy** [comment]
> > is what GNOME uses which render the title bar
> 
> GNOME **does not** draw the titlebar, GNOME forces wayland applications to draw their own titlebars. That's not the case for Windows, macOS, X11 and most of Wayland compositors.

That's what I meant. Sorry for confusion. 

> Avalonia doesn't currently provide its own titlebar and attempting to put one into Window template will break app expectations of full ClientSize being available to them.
> 
> Implementing CSD in a way that's consistent with app expectations on other platforms is doable, but requires work that's currently not planned for the next months.

No just implement that on Linux Wayland. Or even better, just show it on GNOME Wayland (detect Mutter, detect that it's on Wayland and show a simple title bar). If the window has that ExtendToChromeSomethingSomething then just don't show that title bar. On anything else use SSD and let KWin etc. to draw the tite bar for you. 

I mean it's not a clean solution but everyone knows that GNOME won't add SSD support to Mutter in the future and waiting on that promise will halt Avalonia's Wayland support.

You already can make a custom title bar on Avalonia that has the same functions as a normal one. And not just that you also can create window borders that works. I'll make a template for that in the future if anyone wants to use it (and for myself).

**Comment by TheArcaneBrony** [comment]
Couldn't Avalonia report a smaller ClientSize when drawing a title bar?

**Comment by kekekeks** [comment]
>Couldn't Avalonia report a smaller ClientSize when drawing a title bar?

That would mean implementing titlebar in some horrible way instead of just defining it as a normal templated control. Would also still require wl_subsurface for actual content.

There are no hard blockers, just no resources are currently allocated for wayland support.

**Comment by kekekeks** [comment]
I could perhaps have worked on it in my spare time, but it would require enthusiasm and it's hard to have enthusiasm for a tech that still cant run my primary desktop session in 2024 because of various questionable design decisions.

**Comment by amayra** [comment]
> I could perhaps have worked on it in my spare time, but it would require enthusiasm and it's hard to have enthusiasm for a tech that still cant run my primary desktop session in 2024 because of various questionable design decisions.

just ignore Gnome 

they always sabotage any new wayland protocols. Just keep it as drafts for more than 4 years 

just like how game developers ignore mac OS

**Comment by amayra** [comment]
> Would be nice if Wayland support was prioritised, especially given I no longer have a way to reach out for support when things don't work with xwayland.

ryujinx user ?

**Comment by TheArcaneBrony** [comment]
something ryujinx? i dont, no.

**Comment by JakeSays** [comment]
I am actually interested in wayland support only for kiosk style full screen applications. Would there be interest in this level of wayland integration? I think it would be much simpler than full desktop support.

My primary interest is small screen touch only linux devices.

**Comment by Ammar-Sadaoui** [comment]
who still use gnome anyway give us native wayland support and if someone who who use gnome complain we recommend them to upgrade to modern desktop

**Comment by Twig6943** [comment]
> who still use gnome anyway give us native wayland support and if someone who who use gnome complain we recommend them to upgrade to modern desktop

Exactly gnome is known for not implementing/delaying stuff on purpose

**Comment by JakeSays** [comment]
@Ammar-Sadaoui not sure I follow your logic. Those on gnome would still use the X backend so that wouldn't be a problem. I am not for throwing away support for existing systems without ensuring the answer is more than "upgrade your system/desktop/whatever". Having been on the receiving end of that attitude from the likes of google and microsoft, I can assure you it is not pleasant.

**Comment by sandreas** [comment]
> I could perhaps have worked on it in my spare time, but it would require enthusiasm and it's hard to have enthusiasm for a tech that still cant run my primary desktop session in 2024 because of various questionable design decisions.

@kekekeks Did you consider WASM with a thin API layer for input/output (mouse, keyboard, etc.)? Maybe this is totally nuts, but at least Avalonia already has a WASM target.

**Comment by JakeSays** [comment]
@sandreas WASM for what? Not sure how that would facilitate a wayland backend.

**Comment by sandreas** [comment]
> @sandreas WASM for what? Not sure how that would facilitate a wayland backend.

https://www.phoronix.com/news/Greenfield-Compositor-2024

**Comment by JakeSays** [comment]
@sandreas that is not a wayland backend. It is avalonia running in a browser that is apparently optimized for wayland. A wayland backend would not involve any web technologies.

**Comment by sandreas** [comment]
> @sandreas that is not a wayland backend. It is avalonia running in a browser that is apparently optimized for wayland. A wayland backend would not involve any web technologies.

Mmh, i thought the C part of the project is an api layer to bridge wayland to wasm and the js/ts part is the implementation of the compositor...

Sorry when this was wrong.

**Comment by mateli** [comment]
> [@sandreas](https://github.com/sandreas) that is not a wayland backend. It is avalonia running in a browser that is apparently optimized for wayland. A wayland backend would not involve any web technologies.

Except for WebGPU that will likely be used by Skia for rendering on both wayland and X11. Usually trough the wgpu implementation.

**Comment by JakeSays** [comment]
@mateli not sure how that is even remotely relevant. Skia will not be using webgpu through a browser instance to do its rendering, unless it is actually running *in* the browser.

**Comment by sandreas** [comment]
Just to be sure I'm not being misunderstood here: wasm CAN even be run on the Kernel [1, 2] and my thought was to possibly write a thin (!) layer translating the WASM-Version of Avalonia Apps to the according layer of interest (in this case wayland) by adopting parts of existing projects. This way I thought it would not be necessary to have yet another big implementation of a wayland copositor.

Now, that I think this through, I think the required Wasm layer would be even bigger / more complex than a completely new Wayland-Backend. It was kind of a brainstorming thing, how to solve things more easily. Sorry, if that was off topic.

1: https://blog.wasmer.io/running-webassembly-on-the-kernel-8e04761f1d8e

2: https://github.com/wasmerio/kernel-wasm

**Comment by griffi-gh** [comment]
why  create a custom file dialog when there are portals now?

**Comment by mateli** [comment]
> [@mateli](https://github.com/mateli) not sure how that is even remotely relevant. Skia will not be using webgpu through a browser instance to do its rendering, unless it is actually running _in_ the browser.

WebGPU implementations like wgpu are linked directly by many applications and no browsers are needed for that.

**Comment by JakeSays** [comment]
@mateli okay even with wgpu that still isn't a wayland backend. I don't see how that gets us closer to wayland support.

**Comment by mateli** [comment]
Wgpu has support for Wayland. Combine it with DSL or Winit and you probably have all the support you need. Without a single line of Wayland-specific code.

**Comment by Haltroy** [comment]
Wow. It's been few months since I looked back at this and I assume no progress has been made while most major desktop environments and distributions are started to ditch X11. This is frustrating because stuff like fractional scaling, clipboard etc. would not work properly with XWayland apps (basically any Avalonia app). Not to mention how Wayland is more secure than X11 and other stuff.

Look, I don't know what issue that you are having with Wayland session of your desktop environment/compositor of choice but that doesn't mean this shouldn't be implemented. (And what I understand it's GNOME so yes that DE is broken on me as well). 

It's crazy that this issue opened in 2017 and we are at 2025 with no Wayland support (and assuming no progress made to solve it). Aside from the system dialogs since AFAIK Avalonia uses XDG Desktop Portals for that now. I did see some PRs that are now closed but again, closed. This means none of that code got merged. 

My suggestion is to render the decorations on platforms that doesn't have SSD if not implementing it won't crash the app or system but have a weird graphical issue where only content of window is shown (like that DaVinci Resolve bug on GNOME). Feel free to fix my error on that but we can either use internal Avalonia stuff (controls etc.) to draw it ourselves, use libdecor or something else instead or fall back to XWayland (X11) on those platforms (nuclear option). 

When it comes to Wayland, not even .NET libraries are safe from which protocols are implemented or not I guess.

I really like this project and I really want it to shine. I really don't want to use anything else.

**Comment by mateli** [comment]
Let's put it like this. Is directly interfacing with platform specific details the business of a ui toolkit? Or is it the business of a lower level abstraction level. I think it's the latter.

Wine-mono use SDL to for this and I think that this is probably the best solution. Add a layer between SDL and Avalonia if it's needed. But don't introduce platform dependent code unless it's strictly required.

**Comment by JakeSays** [comment]
@mateli Yeah no thank you. The LAST thing I want to deal with is a bunch of unnecessary dependencies like SDL/wgpu/winit/etc. I want a *native* wayland backend, just like we already have a native X backend. Directly interfacing with platform specific details is *exactly* what UI/application toolkits do. That is their point - to abstract away the platform from the application developer.

**Comment by kekekeks** [comment]
>use libdecor or something else instead 

libdecoration and other libs relying on GTK are incompatible with Avalonia's rendering model, sorry

**Comment by M0n7y5** [comment]
Using SDL3 can solve bunch of issues. It exposes multiple render contexts and has support for Wayland out of the box. It can also solve bunch of issues on Android as well. It gives you all the stuff for free. It can be used with Skia without any issues as well. There is a rich history of apps and companies that are using SDL as base for their UI frameworks as well. 
I dont understand what is @JakeSays saying as you don't deal with any dependencies since everything is precompiled as library and shipped as nuget. The usage would not change. 
In terms of NativeAOT, nothing changes also.

**Comment by JakeSays** [comment]
@M0n7y5 Having another native binary, regardless of how it is compiled/packaged/whatever, is one more maintenance burden for both the avalonia developers and users. The user needs to care because they have to make sure that not only do their target environments meet avalonia requirements, but now also SDL3's. So when I port my wayland applications from X64 to ARM, now I need to make sure those additional requirements are met. Avalonia is a very large and complicated framework (such is the nature of any good UI framework), and adding more maintenance complexity is not an advantage.

**Comment by mateli** [comment]
Except when the application already has dependency to SDL3, such as for example quite a few games. Then it isn't even possible to use Avalonia at this point.

**Comment by JakeSays** [comment]
@M0n7y5 I happen to maintain .nuget packages for a set of native binaries for one product. There are TWENTY separate packages to support just windows and linux on all the platforms it needs to support. I fully understand the *real* issue here. But I am done arguing with you, so I will just wait for your super simple, easily maintainable SDL backend.

**Comment by Haltroy** [comment]
The entire SSD vs CSD thing just makes me want to give up. I still don't understand what would happen if we only used SSD-only approach on GNOME etc.?

Aside from "window decoration" thing, what's blocking Wayland support on Avalonia currently?

I'm fine with any backend (SDL, GTK, Qt, Vulkan, OpenGL etc.) as long as it works.

**Comment by JakeSays** [comment]
@Haltroy The only thing blocking a native wayland backend is time and effort. There isn't really anything technically getting in the way - there is just a lot of code that needs to be written.

Take a look at the X11 backend if you want to see what's involved. If you're okay with any backend, then the X11 one works quite well. The issue with it is X11 in general is slowly dying.

My particular interest in wayland is because I wrote a wayland compositor for an embedded system and I would like to use that compositor with avalonia.

**Comment by Haltroy** [comment]
> The only thing blocking a native wayland backend is time and effort. There isn't really anything technically getting in the way - there is just a lot of code that needs to be written.

Oh OK then. I guess you guy managed to solve the "client-side decoration" issue and "how can we port Avalonia to Wayland" issue. I want to know how you guys are approaching on these topics and just Wayland support in general. Hopefully the next PRs won't be closed due to "inactivity" and actually got merged. 

> Take a look at the X11 backend if you want to see what's involved. If you're okay with any backend, then the X11 one works quite well.

I was talking about backends that have native Wayland support. This is just telling someone to "hey do you want a good backend as an alternative to <backend that I don't like> then use <backend that I don't like>".

Again, I'm fine if the backend is SDL or GTK or any other **WAYLAND NATIVE** backend as long as it works. Even a Avalonia native approach would be fine with me. And I think it will be fine with anyone else too as long as it just works.

Personally, I don't use any scaling (but there are people who might do) but the clipboard issues annoy me time to time. Also I'm trying to set-up my system in a way that is Wayland-only. I managed to run my browser, my IDEs etc. to only run on Wayland except Avalonia. Why? Because (in your own words):

> X11 in general is slowly dying.


> My particular interest in wayland is because I wrote a wayland compositor for an embedded system and I would like to use that compositor with avalonia.

Kinda same. I helped (barely) port a CEF wrapper named CefGlue to Linux. It only has code for X11 and I'd like to have a Wayland support to shut that case completely (ever since I was marketed as their "Linux guy", yes they tagged me for Linux-related issues despite me not working in their company) so I can at least sleep knowing that is done at night.

**Comment by JakeSays** [comment]
@Haltroy not sure who you are referring to when you say "you guys" - I am just an avalonia user like you, but I don't want "any" solution - I want the one that is the least burdensome on me and my users, and in spite of what some might think, practical experience has taught me that native dependencies are a huge burden, *especially* in the embedded linux space.

And my condolences on your work with CEF :) I am intimately familiar with working with the chromium source code, having ported chrome to several embedded devices over the years. So I feel your pain there!

**Comment by Haltroy** [comment]
The CEF part wasn't actually hard, I just had to make CEF window to stick to Avalonia and luckily someone else made that part. I just made the rest which is making Linux-specific stuff for the wrapper itself and make a script that downloads builds from Spotify's system and packages it for Nuget. Anyways enough distractions.

I meant it as everyone who's working on Avalonia. I just assumed you work on Avalonia because of the "everything is resolved just coding left" talk. Sorry for that.

That CSD thing still bugs me. Will it just show the window content without any titlebar like that bug I mentioned before or just won't even launch or something else?

I'm just giving up on this. I'm wishing good luck on Avalonia team on Wayland support. This is my last comment on this topic.

**Comment by Ammar-Sadaoui** [comment]
thanks gnome developers for making linux experience better

and we should never forget nvidia too

**Comment by Karlosbubi** [comment]
I only just saw avalonia doesn't support wayland yet and just gonna add my two cents:

>System dialogs (open/save file, folder selection, etc). We need to either still load GTK or have them out-of-proc somehow. The out-of-proc way will cause issues with making dialogs modal to the parent window on X11

As long as linux desktop is targeted (embedded works as long as it includes a desktop) these, this can be handled by the compositor via portals. Which is officially recommended

>Client-side window decorations. Some of the wayland shells are requiring applications to draw window frame, title and buttons, we need to support that somehow

Can be ignored if need be, it's spec compliant (xdg-decorations protocoll) and only one compositor dislikes it, and afaik the recently bowed to the standard and accept it.

**Comment by kekekeks** [comment]
>this can be handled by the compositor via portals

The compositor will need support for `xdg-foreign-unstable-v2` protocol which is "unstable" (there a parttern with every useful protocol being either unstable, staging, compositor-specific or outright rejected), but can work, yes.


We'll likely need 2 separate modes for our wayland support:
1) one that checks if compositor is OK to use by verifying that it has support for various protocols and falling back to X11 if it does not
2) embedded device mode when app author can force-enable it and deal with potential incompatibilities

**Comment by kekekeks** [comment]
As of SDL discussion: we are trying to make sure that out-of-tree platform backends are possible to implement. Feel free to experiment with SDL.

**Comment by Karlosbubi** [comment]
the slow progression of wayland protocols is indeed annoying an might be a slightly increased maintenance burden.
`xdg-decorations` is unstable as well.
At least for Portals i don't see them going anywhere given their adoption and integral importance for stuff like flatpak.

**Comment by Haltroy** [comment]
> The compositor will need support for `xdg-foreign-unstable-v2` protocol which is "unstable" (there a pattern with every useful protocol being either unstable, staging, compositor-specific or outright rejected), but can work, yes.

AFAIK the only compositor that doesn't support [`xdg-foreign-unstable-v2`](https://wayland.app/protocols/xdg-foreign-unstable-v2#compositor-support) (at the moment) is some project named "Hyprland", not a big deal like nobody heard "I use arch btw" meme and yeah I just summoned that guy by talking about his compositor (he's known to be very toxic and yes even he knows it). Things can get dirty here so be ready.

Jokes aside, that guy (notice how I don't mention his name yeah that's a skill I developed after all these years) will probably implement it since he's cracked on stuff like this. Hopefully I'll come back in a few days and overstrike this entire thing ~like this~.

> granted, i'm not completely sure what the exact relation between wayland and portals is, but i'd argue the freedesktop stuff is acceptable dependencies generally (for linux distribution).

And (again AFAIK) both Wayland and XDG Desktop Portals are Freedesktop stuff (well portals is a Flatpak thing which is a Freedesktop thing). Firefox (for dialogs), OBS (for both dialogs and screen capture) and many other software already uses portals (even if its running on distro package instead of Flatpak). XDG Desktop Portals is implemented on all Wayland compositors (by their each own methods like xdg-desktop-portal-kde xdg-desktop-portals-gnome etc.).

> For games, system dialogs are not an option. They require leaving full screen single window mode which completely breaks immersion.

I'd recommend making a dialog box for your own game anyways if you want to keep it as a single window thing. Might be challenging since now you have to deal with file paths, drawing folders etc. but Avalonia's own bindings and a bit of backend code should do the trick on it. 

> Also XDG are not universally available on Linux systems.

It is?? (IDK which systems/distros that don't package it). The only stuff that doesn't support it are either servers, minimal systems or distros that have the package but require manual installation (devs that use Avalonia for their application can use it as a dependency).

I guess everyone (with the exception of few) being on board with SDL (which is nice that's something that Wayland is not capable of doing). I'm fine with falling back to X11 if required protocols aren't implemented on that system.

Alright I'm out once again, just decided to check after getting notified via e-mail.

**Comment by nnbyte** [comment]
Perhaps to inform the discussion on an interesting--**still potential**--development regarding wayland on gnome. 

A [change proposal was just submitted](https://fedoraproject.org/wiki/Changes/WaylandOnlyGNOME), in order to obsolete X11 in favor of Wayland-only GNOME, ideally beginning with Fedora 38, towards the end of this year. 

You might want to keep an eye on this!

**Comment by M0n7y5** [comment]
> Perhaps to inform the discussion on an interesting--**still potential**--development regarding wayland on gnome.
> 
> A [change proposal was just submitted](https://fedoraproject.org/wiki/Changes/WaylandOnlyGNOME), in order to obsolete X11 in favor of Wayland-only GNOME, ideally beginning with Fedora 38, towards the end of this year.
> 
> You might want to keep an eye on this!

This is very interesting. I hope it doesn't break stuff tho.

**Comment by mateli** [comment]
@Haltroy I think that the big problem with game support is to have a uniform pipeline for game content and Avalonia. One way would be to render skia headless and project the result on a 2D or 3D surface. Then the big task would be to map pointer input back to skia.

There are lots of minimalistic Linux distributions without both gnome and XDG. Usually they have X11 or Wayland and then something very minimalistic on top of that. But usually support for SDL/OpenGL/Vulkan etc as games need that. 

Then we also have *BSD where neither Gnome nor XDG are well spread.

**Comment by Haltroy** [comment]
im out stop tagging me

**Comment by Ammar-Sadaoui** [comment]
Any updates on this topic? Just to clarify, I'm only interested in AvaloniaUI because of the Ryujinx emulator — I’ve run into issues caused by XWayland, which is why proper Wayland support matters to me.

**Comment by kekekeks** [comment]
The plan is to have it around 12.0 - 12.2. For 12.0 we are planning to sort out the breaking changes in the core library required for wayland support and maybe enable wayland for embedded systems.

**Comment by jvieira88** [comment]
Hi, just looking for an update on this. 
Looking at some blog posts, it seems like there is a commitment to get this delivered which I really appreciate:
[https://avaloniaui.net/blog/bringing-wayland-support-to-avalonia](url)
[https://avaloniaui.net/blog/three-year-sponsorship-accelerates-avalonia-s-open-source-roadmap](url)

> Avalonia v12 remains on track for Q4

Could we hope for this feature release until the end of the year?

**Comment by jgcodes2020** [comment]
Is #7063 currently the primary blocker (because GNOME?) It seems like other than that, most of the Wayland changes are already good enough.

Before proper CSD is implemented, would it be okay to use something like [libdecor](https://github.com/neonkore/libdecor) to draw fallback decorations?

**Comment by MrJul** [comment]
Wayland support was implemented in https://github.com/AvaloniaUI/Avalonia/pull/21448.
Don't hesitate to play with the nightly builds and give feedback!
